File: train.py
# MIT TASK Y
def find_intersection(a, b):
    return [x for x in a if x in b]

print(find_intersection([1, 2, 3], [3, 2, 0]))



# MIT TASK X
# def count_occurrences(obj, target):
#     count = 0

#     for key, value in obj.items():
#         if key == target:
#             count += 1

#         if isinstance(value, dict):
#             count += count_occurrences(value, target)

#     return count


# MIT TASK
# def chunk_array(arr, size):
#     result = []

#     for i in range(0, len(arr), size):
#         result.append(arr[i:i + size])

#     return result


# print(chunk_array([1, 2, 3, 4, 5], 2))


# MIT TASK V
# def count_char(value):
#     count = {}

#     for char in value:
#         if char in count:
#             count[char]+=1
#         else:
#             count[char] = 1

#     return count
# print(count_char("hello"))


# MIT TASK T
# def merge_sorted_arrays(a, b):
#     return sorted(a + b)

# print(merge_sorted_arrays([0, 3, 4], [4, 6, 2]))


# MIT TASK S
# def missing_number(values):
#     for i in range(len(values) + 1):
#         if i not in values:
#             return i


# print(missing_number([3, 0, 1]))


# MIT TASK R
# def calculate(values):
#     return eval(values)

# print(calculate("3 + 4"))


# def calculate(values):
#     a, b, c = values.split()
#     if b == "+":
#         return int(a) + int(c)


# print(calculate("3 + 4"))


# MIT TASK Q
# def has_property(values, keys):
#     return keys in values


# print(has_property({"age": "BMW", "name": "BMW"}, "name"))

# def has_property(values, keys):
#     for cuurent_key in values:
#         if cuurent_key == keys:
#             return True

#     return False


# print(has_property({"age": "BMW", "name": "BMW"}, "name"))


# MIT TASK P
# def object_to_array(keys):
#     return list(keys.items())

# print(object_to_array({"a": 10, "b": 20}))
# def object_to_array(values):
#     result = []
#     for key in values:
#         result.append([key, values[key]])

#     return result


# print(object_to_array({"a": 10, "b": 20}))


# MIT TASK O
# def calculate(values):
#     total = 0

#     for value in values:
#         if isinstance(value, (int, float)):
#             total += value

#     return total


# print(calculate([10, "10", {"son": 10}, "True", 35]))


# MIT TASK N
# def palindrom_check(value):
#     return value == value[::-1]
# print(palindrom_check("dad"))


# MIT TASK M
# def get_square(values):
#     result = []
#     for value in values:
#         result.append({
#             "number": value,
#             "square": value ** 2
#         })
#     return result


# print(get_square([1, 2, 4]))


# def reverse_value(value):
#     words = value.split(" ")
#     result = []
#     for word in words:
#         result.append(word[::-1])

#     return " ".join(result)


# print(reverse_value("we like coding"))

File: train.ts
//MIT TASK S
function missingNumber(values: any) {
  for (let i = 0; i <= values.length; i++) {
    if (!values.includes(i)) {
      return i;
    }
  }
}

console.log(missingNumber([3, 0, 1]));

// <MIT TASK R
// function calculate(values: any): any {
//   return eval(values)

// }
// console.log(calculate("6 + 4"));

// function calculate(values: any): any {
//   const [a, b, c] = values.split(" ");
//   if (b === "+") {
//     return Number(a) + Number(c);
//   }
// }
// console.log(calculate("6 + 4"));
//MIT TASK Q

// function hasProperty(values: any, keys: string): any {
//   for (const key in values) {
//     if (key === keys) {
//       return true;
//     }
//   }
//   return false;
// }
// console.log(hasProperty({ age: "BMW", year: "BMW" }, "name"));

// function hasProperty(values: any, keys: string): any {
//   return keys in values
// }
// console.log(hasProperty({ age: "BMW", name: "BMW" }, "name"));

// function hasProperty(values: any, keys: string): boolean {
//   return values.hasOwnProperty(keys);
// }
// console.log(hasProperty({ age: "BMW", year: "BMW" }, "name"));

//MIT TASK P
//1)// function objectToArray(keys: any) {
//   let nest = [];
//   for (const key in keys) {
//     nest.push([key, keys[key]]);
//   }
//   return nest;
// }
// console.log(objectToArray({ a: 10, b: 20 }));

//2)
// function objectToArray(keys: any) {
//   return Object.entries(keys);
// }

// console.log(objectToArray({ a: 10, b: 20 }));

//MIT TASK O
// function calculate(values: any) {
//   let sum: any = 0;
//   for (let value of values) {
//     if (typeof value === "number") {
//       sum += value;
//     }
//   }
//   return sum;
// }
// console.log(calculate([10, "10", { son: 10 }, true, 35]));

//MIT TASK N
// function palindrom(value: string): boolean {
//   return value.split("").join("") === value.split("").reverse().join("");
// }
// console.log(palindrom("dad"));

//MIT TASK M
// function getSquareNumbers(values: number[]) {
//   return values.map((value: number) => {
//     return {
//       number: value,
//       square: value * value,
//     };
//   });
// }
// console.log(getSquareNumbers([1, 2, 3]));

// function getSquareNumbers(values: number[]) {
//   const result = [];
//   for (const value of values) {
//     result.push({
//       number: value,
//       square: value * value,
//     });
//   }
//   return result;
// }
// console.log(getSquareNumbers([1, 2, 3]));
//MIT TASK L

// function reversevalue(values: string) {
//   return values
//     .split(" ")
//     .map((word: string) => {
//       return word.split("").reverse().join("");
//     })
//     .join(" ");
// }
// console.log(reversevalue("we like coding"));

// Architectural pattern: MVC, Dependency Injection, MVP
//MVC = model view controller
//Design pattern: Middleware, Decorator

// CLUSTER => Database => Collection => Document

/* Project Standards:
 - logging standards 
 - naming stardards 
    functions, methods, variables => Camel     
   class => PASCAL                       
    folder => KEBAB
    css => SNAKE 


#  -ERROR handling  
# */

/*
# (traditional)APi
# RestApi
# GraphQL APi 
# */

/*
 Traditional Frontend Development => BSSR (Admin) => EJS

 Modern frontend Development   =>  SPA(User) => REACT 

*/

/**COOKIES
 * 
 request join
 self destroy

 */
